pkg/store/storepb/custom.go

```diff
@@ -381,31 +384,112 @@ func PromMatchersToMatchers(ms ...*labels.Matcher) ([]LabelMatcher, error) {
+	case LabelMatcher_RE:
+		t = labels.MatchRegexp
+	case LabelMatcher_NRE:
+		t = labels.MatchNotRegexp
+	default:
+		return nil, errors.Errorf("unrecognized label matcher type %d", m.Type)
+	}
+	pm, err := labels.NewMatcher(t, m.Name, m.Value)
+	if err != nil {
+		return nil, err
+	}
+	return pm, nil
+}
+
 // MatchersToPromMatchers returns Prometheus matchers from proto matchers.
 // NOTE: It allocates memory.
 func MatchersToPromMatchers(ms ...LabelMatcher) ([]*labels.Matcher, error) {
 	res := make([]*labels.Matcher, 0, len(ms))
 	for _, m := range ms {
-		var t labels.MatchType
+		m, err := MatcherToPromMatcher(m)
+		if err != nil {
+			return nil, err
+		}
+		res = append(res, m)
+	}
+	return res, nil
+}
 
-		switch m.Type {
-		case LabelMatcher_EQ:
-			t = labels.MatchEqual
-		case LabelMatcher_NEQ:
-			t = labels.MatchNotEqual
-		case LabelMatcher_RE:
-			t = labels.MatchRegexp
-		case LabelMatcher_NRE:
-			t = labels.MatchNotRegexp
-		default:
-			return nil, errors.Errorf("unrecognized label matcher type %d", m.Type)
+type MatcherConverter struct {
+	cache         *cache.TwoQueueCache[LabelMatcher, *labels.Matcher]
+	cacheCapacity int
+	metrics       *matcherConverterMetrics
+}
+
+type matcherConverterMetrics struct {
+	cacheTotalCount prometheus.Counter
+	cacheHitCount   prometheus.Counter
+	cacheSizeGauge  prometheus.Gauge
+}
+
+func newMatcherConverterMetrics(reg prometheus.Registerer) *matcherConverterMetrics {
+	var m matcherConverterMetrics
+
+	m.cacheTotalCount = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_store_matcher_converter_cache_total",
+		Help: "Total number of cache access.",
+	})
+	m.cacheHitCount = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_store_matcher_converter_cache_hit_total",
+		Help: "Total number of cache hits.",
+	})
+	m.cacheSizeGauge = promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name: "thanos_store_matcher_converter_cache_size",
+		Help: "Current size of the cache.",
+	})
+
+	return &m
+}
+
+// NewMatcherConverter creates a new MatcherConverter with given capacity.
+func NewMatcherConverter(cacheCapacity int, reg prometheus.Registerer) (*MatcherConverter, error) {
+	c, err := cache.New2Q[LabelMatcher, *labels.Matcher](cacheCapacity)
+	if err != nil {
+		return nil, err
+	}
+	metrics := newMatcherConverterMetrics(reg)
+	return &MatcherConverter{cache: c, cacheCapacity: cacheCapacity, metrics: metrics}, nil
+}
+
+// MatchersToPromMatchers converts proto label matchers to Prometheus label matchers. It caches regex conversions.
+func (c *MatcherConverter) MatchersToPromMatchers(ms ...LabelMatcher) ([]*labels.Matcher, error) {
+	res := make([]*labels.Matcher, 0, len(ms))
+	for _, m := range ms {
+		if m.Type == LabelMatcher_EQ || m.Type == LabelMatcher_NEQ {
+			// EQ and NEQ are very cheap, so we don't cache them.
+			pm, err := MatcherToPromMatcher(m)
+			if err != nil {
+				return nil, err
+			}
+			res = append(res, pm)
+			continue
+		}
```

Comment:
Suggest caching only REGEX operators to be future-proof (e.g., another non-regex operator is added)
```
if m.Type != LabelMatcher_RE && m.Type != LabelMatcher_NRE {
```